wordpress-mobile/WordPress-iOS
# QR Login: Add Networking

| 📚 Project: #18594 | 🔗  Depends on: #18752 | 🤝 WPKit Related PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/513|
|:---:|:---:|:---:|

### Requirements 
Prior to testing the PR you will need to:
- Follow the testing steps in the related Calypso PR: 
    - https://github.com/Automattic/wp-calypso/pull/63257
- Make sure your **real device** is sandboxed: pbArwn-24D-p2
- Make sure your sandbox is pointed to D79766-code 

### Description
- Implements the QR Login network requests
- Adds some extra error states to the verify view
- Adds a class `QRLoginURLParser` that parses the real QR login token URLs
- Adds a way to "scan again" from the verify view 
- Updates the verify view to show the requesting users Browser, and approximate location

### Notes
- In a later PR a "cancel" button is added to the error states to allow the user to leave the flow without needing to tap scan again
- In a later PR the red subtitle label on the "You're logged in" state is changed to gray

### Demo Video

https://user-images.githubusercontent.com/793774/170776717-9e81c6e5-7d18-47b1-a40c-5cbf6bfd2f20.mov

### Screenshots
| Are you sure? | Success! | No connection | Invalid Data | Expired Token | 
|:---:|:---:|:---:|:---:|:---:|
|![01 - are you sure](https://user-images.githubusercontent.com/793774/170783226-434233da-8223-403b-b9d3-add118eae26c.PNG)|![01 - logged in](https://user-images.githubusercontent.com/793774/170783224-1d262a86-4dc4-428b-a358-7974ed422964.PNG)|![01 - no connection](https://user-images.githubusercontent.com/793774/170783227-ac9423ce-e64e-4da1-ad9d-a5205f9becdd.PNG)|![01 - invalid data](https://user-images.githubusercontent.com/793774/170783222-ad95377a-2ed2-4c92-a951-7632488f3a34.PNG)|![01 - expired](https://user-images.githubusercontent.com/793774/170783219-5205afca-3e3e-478d-8854-e4075e071909.PNG)|








### To test:

#### Scanning a token with invalid data
![invalid-qrlogin-code](https://user-images.githubusercontent.com/793774/170778949-f8b47a11-ce27-4db0-87f5-85f5be09ec07.png)

1. Launch the app
2. Navigate to My Site > Me > Scan Login Code
3. Scan the QR code image from above
4. 👁️ Verify you see the "Could not validate" error view
5. 👁️ Verify tapping Scan again takes you to the scanning view again

#### Scanning a valid but expired token
![qrlogin-expired](https://user-images.githubusercontent.com/793774/170781480-d8fc1f46-6052-4e67-94e7-257b96689562.png)

1. Launch the app
2. Navigate to My Site > Me > Scan Login Code
3. Scan the QR code image from above
4. 👁️ Verify you see the "Expired log in code" error view
5. 👁️ Verify tapping Scan again takes you to the scanning view again

#### Scanning a valid token
1. Launch the app
2. Navigate to My Site > Me > Scan Login Code
3. Scan the QR Code from the related Calypso PR (See above)
4. 👁️ Verify you see a "Are you trying to login" message that displays your web browser and approximate location
    5. "Are you trying to login to Chrome near Los Angeles, California?"
6. 👁️ Verify tapping Cancel exits out of the flow and nothing else happens
7. Scan the valid code again
8. Tap the 'Yes, log me in' button
9. 👁️ Verify you see a loading indicator with the current message faded out
10. 👁️ Verify after a second you see the "You're logged in!" message
11. 👁️ Verify tapping Dismiss exits the flow
12. 👁️ Verify you are logged in on your web browser

#### Testing no internet error
1. Launch the app
3. Navigate to My Site > Me > Scan Login Code
4. Turn on Airplane mode
5. Scan the QR Code from the related Calypso PR (See above)
6. 👁️ Verify you see a "No connection" error state
7. 👁️ Tapping Scan again brings you back to the scanning view
8. Enable your internet and scan again
9.  👁️ Verify the scanning works correctly 

### Regression Notes
1. Potential unintended areas of impact
None

13. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

14. What automated tests I added (or what prevented me from doing so)
None

PR submission checklist:
